accord-core/src/main/java/accord/coordinate/FetchData.java

```diff
@@ -415,14 +444,92 @@ public Void apply(SafeCommandStore safeStore)
                     break;
             }
 
+            return updateDurability(safeStore, safeCommand);
+        }
+
+        // if can only propagate Truncated, we might be stale; try to upgrade the
+        private Known applyOrUpgradeTruncated(SafeCommandStore safeStore, SafeCommand safeCommand, Command command)
+        {
+            // if our peers have truncated this command, then either:
+            // 1) we have already applied it locally; 2) the command doesn't apply locally; 3) we are stale; or 4) the command is invalidated
+            if (command.saveStatus().isUninitialised())
+                return null; // TODO (expected): maybe stale?
+
+            if (command.hasBeen(PreApplied))
+            {
+                updateDurability(safeStore, safeCommand);
+                return null;
+            }
+
+            if (Infer.safeToCleanup(safeStore, command, route, full.executeAt))
+            {
+                Commands.setErased(safeStore, safeCommand);
+                return null;
+            }
+
+            // we're now at least partially stale, but let's first see if we can progress this shard, or we can do so in part
+            Timestamp executeAt = command.executeAtIfKnown(full.executeAtIfKnown());
+            if (executeAt == null)
+            {
+                // we don't even know the execution time, so we cannot possibly proceed besides erasing the command state and marking ourselves stale
+                // TODO (required): we could in principle be stale for future epochs we haven't witnessed yet. Ensure up to date epochs before finalising this application.
+                safeStore.commandStore().markShardStale(safeStore, txnId, route.participants().toRanges(), false);
+                Commands.setErased(safeStore, safeCommand);
+                return null;
+            }
+
+            Ranges executeRanges = safeStore.ranges().allBetween(txnId, executeAt);
+            executeRanges = safeStore.commandStore().redundantBefore().expectToExecute(txnId, executeAt, executeRanges);
+
+            if (executeRanges.isEmpty())
+            {
+                // TODO (expected): we might prefer to adopt Redundant status, and permit ourselves to later accept the result of the execution and/or definition
+                Commands.setTruncatedApply(safeStore, safeCommand, route);
+                return null;
+            }
+
+            Known required = PreApplied.minKnown;
```

Comment:
Thanks, formalising my reply made me realise that this is perhaps wrong. I had failed to distinguish between when we are truncated for the whole node, and for the command store in question. `applyOrUpgradeTruncated` currently assumes the command is truncated for the command store's range, when in fact it may only be truncated for the node as a whole. 

The basic premise was that if we are truncated then we should have been durably `PreApplied`, so if we cannot obtain enough information to pre-apply locally we must be stale. It is however permitted for one shard to have fully applied and to partially truncate (i.e. TruncatedApplyWithOutcome) while another shard has not even recorded its `committedDeps`, e.g. in the case of fast-path commit that only applies on one shard. So this is probably a bug.